File: examples/orchestration/node/src/02_streaming_events.ts
/**
 * Streaming Events
 *
 * Demonstrates real-time event streaming by polling nextEvent()
 * until the execution completes (returns null).
 */

import * as fs from "fs";
import * as path from "path";
import * as os from "os";

const native = require("../../../crates/cleargate-node/cleargate.node");

function main() {
  const tmpDir = fs.mkdtempSync(path.join(os.tmpdir(), "cleargate-"));
  const flowDir = path.join(tmpDir, "flows");
  const runDir = path.join(tmpDir, "runs");

  const flowDest = path.join(flowDir, "echo-flow");
  fs.mkdirSync(flowDest, { recursive: true });
  fs.copyFileSync(
    path.join(__dirname, "../../flows/echo-flow.json"),
    path.join(flowDest, "flow.json"),
  );

  try {
    const builder = new native.NapiEngineBuilder();
    builder.flowStorePath(flowDir);
    builder.runStorePath(runDir);
    builder.crashRecovery(false);
    const engine = builder.build();

    console.log("Executing echo-flow with streaming events...\n");
    const handle = engine.execute("echo-flow", { message: "Stream test" });

    const start = Date.now();
    let event;
    while ((event = handle.nextEvent()) !== null) {
      const elapsed = ((Date.now() - start) / 1000).toFixed(3);
      const type = Object.keys(event)[0];
      const payload = event[type];
      console.log(`  [${elapsed}s] ${type}`);
      if (payload?.node_instance_id) {
        console.log(`           node: ${payload.node_instance_id}`);
      }
    }

    console.log(
      `\nAll events received in ${((Date.now() - start) / 1000).toFixed(3)}s`,
    );
    engine.shutdown();
  } finally {
    fs.rmSync(tmpDir, { recursive: true, force: true });
  }
}

main();

File: examples/orchestration/python/02_streaming_events.py
#!/usr/bin/env python3
"""Streaming Events

Demonstrates real-time event streaming using the iterator protocol:
  for event in handle:
      process(event)
"""

import os
import shutil
import tempfile
import time

from cleargate import EngineBuilder


def main():
    flow_src = os.path.join(os.path.dirname(__file__), "..", "flows", "echo-flow.json")
    with tempfile.TemporaryDirectory() as tmpdir:
        flow_dir = os.path.join(tmpdir, "flows")
        run_dir = os.path.join(tmpdir, "runs")

        flow_dest = os.path.join(flow_dir, "echo-flow")
        os.makedirs(flow_dest, exist_ok=True)
        shutil.copy(flow_src, os.path.join(flow_dest, "flow.json"))

        builder = EngineBuilder()
        builder.flow_store_path(flow_dir)
        builder.run_store_path(run_dir)
        builder.crash_recovery(False)
        engine = builder.build()

        print("Executing echo-flow with streaming events...\n")
        handle = engine.execute("echo-flow", {"message": "Stream test"})

        start = time.monotonic()
        for event in handle:
            elapsed = time.monotonic() - start
            # Each event is a dict with a single key (the event variant)
            for event_type, payload in event.items():
                print(f"  [{elapsed:.3f}s] {event_type}")
                if isinstance(payload, dict) and "node_instance_id" in payload:
                    print(f"           node: {payload['node_instance_id']}")

        print(f"\nAll events received in {time.monotonic() - start:.3f}s")
        engine.shutdown()


if __name__ == "__main__":
    main()
